src/WebJobs.Script/Diagnostics/ConditionalTraceWriter.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using Microsoft.Azure.WebJobs.Host;
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    public sealed class ConditionalTraceWriter : TraceWriter
+    {
+        private readonly Func<TraceEvent, bool> _predicate;
+        private readonly FailedConditionTraceBehavior _failedConditionBehavior;
+
+        public ConditionalTraceWriter(TraceWriter innerWriter, Func<TraceEvent, bool> predicate, FailedConditionTraceBehavior failedConditionBehavior = FailedConditionTraceBehavior.TraceAsVerbose) 
+            : base(innerWriter?.Level ?? TraceLevel.Off)
+        {
+            if (innerWriter == null)
+            {
+                throw new ArgumentNullException(nameof(innerWriter));
+            }
+
+            if (predicate == null)
+            {
+                throw new ArgumentNullException(nameof(predicate));
+            }
+
+            InnerWriter = innerWriter;
+            _predicate = predicate;
+            _failedConditionBehavior = failedConditionBehavior;
+        }
+
+        public TraceWriter InnerWriter { get; }
+
+        public override void Trace(TraceEvent traceEvent)
+        {
+            if (_predicate(traceEvent))
+            {
+                InnerWriter.Trace(traceEvent);
+            }
+            else if (_failedConditionBehavior == FailedConditionTraceBehavior.TraceAsVerbose)
+            {
+                var eventClone = CloneEvent(traceEvent);
+
+                // To improve troubleshooting, instead of completely suppressing the event, we'll downgrade
+                // it to 'Verbose' and add a prefix to the message.
```

Comment:
What if the message is already Verbose and you only want it to go on primary? Looks like this means you can't filter out Verbose?

I'm wondering if we shouldn't just use the "debug mode" work I'm going to be following up with to drive whether we force these to go through. I'm going to be forcing other diagnostics/log level stuff based on whether the debug mode is on (or has been forced on). However, I can consider that rework in my PR if you wish.